### ISSUE TITLE
Upgrade error on role switch for view-only

### ISSUE THREAD

**Issue by mjnagel** [kiali/kiali#5695]
This issue I believe is a result of https://github.com/kiali/kiali/issues/5611 and the associated PR.

When upgrading my existing Kiali install to 1.61.0 I am seeing the following error:

```
message: 'Failed to patch object: b''{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"ClusterRoleBinding.rbac.authorization.k8s.io \\"kiali\\" is invalid: roleRef: Invalid value: rbac.RoleRef{APIGroup:\\"rbac.authorization.k8s.io\\", Kind:\\"ClusterRole\\", Name:\\"kiali-viewer\\"}: cannot change roleRef","reason":"Invalid","details":{"name":"kiali","group":"rbac.authorization.k8s.io","kind":"ClusterRoleBinding","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: rbac.RoleRef{APIGroup:\\"rbac.authorization.k8s.io\\", Kind:\\"ClusterRole\\", Name:\\"kiali-viewer\\"}: cannot change roleRef","field":"roleRef"}]},"code":422}\n'''
```

As mentioned here the roleRef is immutable, causing this error/invalid I believe - https://kubernetes.io/docs/reference/access-authn-authz/rbac/#clusterrolebinding-example

> After you create a binding, you cannot change the Role or ClusterRole that it refers to. If you try to change a binding's roleRef, you get a validation error. If you do want to change the roleRef for a binding, you need to remove the binding object and create a replacement.

I didn't see any open issues or release notes mentioning this upgrade issue - is this something we could account for in the operator (deleting the current binding, then making a replacement)?

**Comment by mjnagel** [comment]
For additional context on the setup - I was upgrading from 1.60.0 to 1.61.0 with openid as my auth strategy. The linked issue/PR adjusted the default role in my scenario to be a read only one if I understand correctly, which causes this error on upgrade.

**Comment by jmazzitelli** [comment]
Yes, this is one of the few times where a straight upgrade will not work because Kubernetes will not let the Kiali operator modify certain portions of the rolebinding resource.

Workaround: You must delete the old Kiali CR entirely before upgrading the operator.  Then re-create your Kiali CR.

**Comment by mjnagel** [comment]
> Workaround: You must delete the old Kiali CR entirely before upgrading the operator. Then re-create your Kiali CR.

Admittedly I haven't thought this fully through but could we add a condition [here](https://github.com/kiali/kiali-operator/blob/master/roles/default/kiali-deploy/tasks/main.yml#L788) to check for this upgrade (pre-1.61, etc) and then run that delete. That would provide a clean upgrade process, although I'm not sure on the proper condition.

I'd be happy to try and work a change for this but might need some pointers and perspective on whether this is feasible. 

If it's not feasible does deleting just the role/clusterrole work as a manual upgrade process rather than the full Kiali CR? The manual upgrade step would be great to add in a changelog/release notes somewhere if that ends up being the recommendation.

**Comment by jmazzitelli** [comment]
Here's the operator code in question - now, this all should have deleted the role prior to creating it, however I suspect the issue is because the old 1.60 role was read-write and no way for the operator today to know that (it assumes current behavior which is default read-only).

https://github.com/kiali/kiali-operator/blob/v1.61.0/roles/default/kiali-deploy/tasks/main.yml#L782-L802

Probably need to figure out a good "when" clause item here to fix it for this edge-case.

**Comment by jmazzitelli** [comment]
It could be the easiest thing to do is just document that before you upgrade, set `spec.deployment.view_only_mode=true` in the Kiali CR. I think (not 100% sure) that would work around the problem in the easiest way.

**Comment by jmazzitelli** [comment]
how I replicated:

1. Install minikube with Istio installed in `istio-system`
2. Install v1.60 operator via `helm install --namespace kiali-operator --create-namespace --version 1.60.0 kiali-operator kiali/kiali-operator`
3. Create a Kiali CR file `/tmp/kiali-cr.yaml` that looks like this:
```yaml
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali
  namespace: istio-system
spec:
  auth:
    strategy: token
  deployment:
    accessible_namespaces: [ "**" ]
```
4. Install Kiali Server by `kubectl apply -f /tmp/kiali-cr.yaml`
5. Upgrade to v1.61 operator via `helm install --namespace kiali-operator --create-namespace --version 1.61.0 kiali-operator kiali/kiali-operator`

If you look at the status field of the Kiali CR (`kubectl get kiali kiali -n istio-system -o yaml`) you will see this:

```
    message: 'Failed to patch object: b''{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"ClusterRoleBinding.rbac.authorization.k8s.io
      \\"kiali\\" is invalid: roleRef: Invalid value: rbac.RoleRef{APIGroup:\\"rbac.authorization.k8s.io\\",
      Kind:\\"ClusterRole\\", Name:\\"kiali-viewer\\"}: cannot change roleRef","reason":"Invalid","details":{"name":"kiali","group":"rbac.authorization.k8s.io","kind":"ClusterRoleBinding","causes":[{"reason":"FieldValueInvalid","message":"Invalid
      value: rbac.RoleRef{APIGroup:\\"rbac.authorization.k8s.io\\", Kind:\\"ClusterRole\\",
      Name:\\"kiali-viewer\\"}: cannot change roleRef","field":"roleRef"}]},"code":422}\n'''
```

**Comment by jmazzitelli** [comment]
[PR](https://github.com/kiali/kiali-operator/pull/600) is submitted that fixes this bug. Test procedures are in the PR. This fix will be in the next 1.63 release barring any problems detected in that PR.
